File: assgn1/Percolation.java
/*
 * Author: Param Singh
 * Date: 25/06/2015
 */
public class Percolation {
    // boolean grid for all sites
    private boolean[][] id;

    private WeightedQuickUnionUF qf;
    private int size;
    private int top;
    private int bottom;

    public Percolation(int N) {
        // N cannot be less than or equal to zero
        if (N <= 0) {
            throw new IllegalArgumentException("N should be greater than 0");
        }

        id = new boolean[N][N];

        // quickfind structure for storing connections
        // contains n*n + 2 values because I am using
        // 2 extra spots to see if connected to the top
        // and bottom
        qf = new WeightedQuickUnionUF(N*N + 2);

        size = N;
        top = size * size;
        bottom = size * size + 1;

        int i, j;
        for (i = 0; i < N; i++)
            for (j = 0; j < N; j++)
                // All sites are blocked by default
                id[i][j] = false;
    }

    private void validate(int x, int y) {
        if (x <= 0 || y > size)
            throw new IndexOutOfBoundsException();
    }

    private int convert(int i, int j) {
        return i*size + j;
    }

    public void open(int i, int j) {
        validate(i, j);
        int x = i-1;
        int y = j-1;
        id[x][y] = true;
        int current = convert(x, y);

        // union the cells above and below
        if (x == 0) {
            if (isOpen(x+2, y+1))
                qf.union(current, convert(x+1, y));
            qf.union(current, top);
        }
        else if (x == size-1) {
            if (isOpen(x, y+1))
                qf.union(current, convert(x-1, y));
            qf.union(current, bottom);
        }
        else {
            if (isOpen(x, y+1))
                qf.union(current, convert(x-1, y));
            if (isOpen(x+2, y+1))
                qf.union(current, convert(x+1, y));
        }

        // union the cells left and right
        if (y != size-1 && y != 0) {
            if (isOpen(x+1, y))
                qf.union(current, convert(x, y-1));
            if (isOpen(x+1, y+2))
                qf.union(current, convert(x, y+1));
        }
        else if (y == size - 1) {
            if (isOpen(x+1, y))
                qf.union(current, convert(x, y-1));
        }
        else if (y == 0) {
            if (isOpen(x+1, y+2))
                qf.union(current, convert(x, y+1));
        }
    }

    public boolean isOpen(int i, int j) {
        validate(i, j);
        return id[i-1][j-1];
    }

    public boolean isFull(int i, int j) {
        validate(i, j);
        return qf.connected(convert(i-1, j-1), top);
    }

    public boolean percolates() {
        return qf.connected(top, bottom);
    }

    public static void main(String[] args) {
        System.out.println("Enter the size of the grid");
        int size = StdIn.readInt();
        Percolation p = new Percolation(size);
        while (true) {
            System.out.println("Enter a cell to open");
            int i = StdIn.readInt();
            int j = StdIn.readInt();
            p.open(i, j);
            if (p.percolates())
                System.out.println("Percolates");
            else
                System.out.println("Does not percolate");
        }
    }
}

File: python/union-find/percolation.py
from weightedQuickUnionUF import WeightedQuickUnionUF

class Percolation:
    def __init__(self, n):
        if n <= 0:
            raise ValueError("n should be greater than 0")

        self._grid = [[False for _ in range(n)] for _ in range(n)]
        self._size = n

        # a quick find data structure containing sites for
        # virtual tops and virtual bottoms
        self._qf = WeightedQuickUnionUF(self._size*self._size + 2)

        # a quick find data structure only containing a site for
        # virtual top (to eliminate backwash)
        self._nf = WeightedQuickUnionUF(self._size * self._size + 1)

        self._top = self._size * self._size
        self._bottom = self._top + 1

    def _validate(self, i , j):
        """Check if indexes passed are valid, if invalid, throw an IndexError"""
        if i <= 0 or i > self._size or j <= 0 or j > self._size:
            raise IndexError(str(i) + ", " + str(j))

    def open(self, i, j):
        """ Open site (row i, column j) if it is not open already"""
        self._validate(i, j)

        x = i - 1
        y = j - 1
        if self.is_open(i, j):
            return

        self._grid[x][y] = True

        # corner case
        # if grid is 1 x 1
        if self._size == 1:
            self._qf.union(self._convert(x, y), self._top)
            self._qf.union(self._convert(x, y), self._bottom)
            self._nf.union(self._convert(x, y), self._top)
            return

        if x == 0:
            # only check the site below
            if self.is_open(i+1, j):
                self._qf.union(self._convert(x, y), self._convert(x+1, y))
                self._nf.union(self._convert(x, y), self._convert(x+1, y))
            self._qf.union(self._convert(x, y), self._top)

        elif x == self._size - 1:
            # only check the site above
            if self.is_open(i-1, j):
                self._qf.union(self._convert(x, y), self._convert(x-1, y))
                self._nf.union(self._convert(x, y), self._convert(x-1, y))
            self._qf.union(self._convert(x, y), self._bottom)

        else:
            # check both sites above and below
            if self.is_open(i-1, j):
                self._qf.union(self._convert(x, y), self._convert(x-1, y))
                self._nf.union(self._convert(x, y), self._convert(x-1, y))

            if self.is_open(i+1, j):
                self._qf.union(self._convert(x, y), self._convert(x+1, y))
                self._nf.union(self._convert(x, y), self._convert(x+1, y))

        if y == 0:
            # check only the right site
            if self.is_open(i, j+1):
                self._qf.union(self._convert(x, y), self._convert(x, y+1))
                self._nf.union(self._convert(x, y), self._convert(x, y+1))
        elif y == self._size - 1:
            if self.is_open(i, j-1):
                self._qf.union(self._convert(x, y), self._convert(x, y-1))
                self._nf.union(self._convert(x, y), self._convert(x, y-1))
        else:
            if self.is_open(i, j-1):
                self._qf.union(self._convert(x, y), self._convert(x, y-1))
                self._nf.union(self._convert(x, y), self._convert(x, y-1))

            if self.is_open(i, j+1):
                self._qf.union(self._convert(x, y), self._convert(x, y+1))
                self._nf.union(self._convert(x, y), self._convert(x, y+1))


    def _convert(self, i, j):
       return self._size * i + j

    def is_open(self, i, j):
        self._validate(i, j)

        return self._grid[i-1][j-1]

    def is_full(self, i, j):
        self._validate(i, j)

        return self._nf.connected(self._convert(i-1, j-1), self._top)

    def percolates(self):
        return self._qf.connected(self._top, self._bottom)

if __name__ == '__main__':
    n = int(raw_input("Enter the size: "))
    p = Percolation(n)
    print "Enter 'quit' to quit"
    i = ''
    while i != 'quit':
        x, y = map(int, raw_input().split())
        if not p.is_open(x, y):
            p.open(x, y)
        if p.percolates():
            print "Percolates"
        else:
            print "Does not percolate"
